console/src/storage/store.rs

```diff
@@ -0,0 +1,143 @@
+/// Currently, the console doesn't do such kind of memory optimization.
+#[derive(Debug, Default)]
+pub struct Store {
+    events: Vec<EventEntry>,
+    spans: Vec<Span>,
+
+    updated: bool,
+    id_counter: usize,
+    id_map: HashMap<u64, InternalId>,
+}
+
+impl Store {
+    pub fn new() -> Store {
+        Store::default()
+    }
+
+    pub fn updated(&self) -> bool {
+        self.updated
+    }
+
+    pub fn clear(&mut self) {
+        self.updated = false;
+    }
+
+    pub fn events(&self) -> &[EventEntry] {
+        &self.events
+    }
+
+    pub fn spans(&self) -> &[Span] {
+        &self.spans
+    }
+}
+
+/// See `Store` documentation
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct InternalId(usize);
+
+#[derive(Debug)]
+pub struct Span {
+    id: InternalId,
+    span: NewSpan,
+
+    records: Vec<Record>,
+    follows: Vec<SpanId>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct EventEntry {
+    pub span: Option<InternalId>,
+    pub event: Event,
+}
+
+impl EventEntry {
+    pub fn level(&self) -> Option<Level> {
+        Level::from_i32(self.event.metadata.as_ref()?.level)
+    }
+}
+
+/// Convenience Wrapper around `Arc<Mutex<Store>>`
+#[derive(Clone, Default)]
+pub struct StoreHandle(pub Arc<Mutex<Store>>);
+
+impl StoreHandle {
+    pub fn new() -> StoreHandle {
+        StoreHandle::default()
+    }
+
+    /// Locks and updates the underlying `Store`
+    pub fn handle(&self, variant: Variant) {
+        let mut store = self.0.lock().unwrap();
+        match variant {
+            Variant::NewSpan(span) => store.new_span(span),
+            Variant::Record(record) => store.record(record),
+            Variant::Follows(follows) => store.record_follows_from(follows),
+            Variant::Event(event) => store.event(event),
+        }
+    }
+}
+
+impl Store {
+    fn new_span(&mut self, span: NewSpan) {
+        // Update id mapping for span, see `Store` documentation
+        self.id_map.insert(
+            span.span
+                .as_ref()
+                .expect("BUG: No id assined to NewSpan")
+                .id,
+            InternalId(self.id_counter),
+        );
+
+        self.spans.push(Span {
+            id: InternalId(self.id_counter),
+            span,
+            records: vec![],
+            follows: vec![],
+        });
+        self.id_counter += 1;
+    }
+
+    fn record_follows_from(&mut self, follows: RecordFollowsFrom) {
+        let span = self.id_map[&follows
+            .span
+            .as_ref()
+            .expect("BUG: No id set on follows.span")
+            .id];
+        self.spans[span.0]
+            .follows
+            .push(follows.follows.expect("BUG: No id set on follows.follows"));
+    }
+
+    fn record(&mut self, record: Record) {
+        self.updated = true;
+        let span = self.id_map[&record
+            .span
+            .as_ref()
+            .expect("BUG: No id set on follows.span")
```

Comment:
Nit: I think you meant to say 
```suggestion
            .expect("BUG: No id set on record.span")
```